File: agent/dotnet/src/Perper.Model/AsyncLocals.cs
using System;
using System.Threading;
using System.Threading.Tasks;
using Perper.Protocol.Service;

namespace Perper.Model
{
    public static class AsyncLocals
    {
        private static AsyncLocal<(CacheService, NotificationService)> _connection = new AsyncLocal<(CacheService, NotificationService)>();
        private static AsyncLocal<string> _instance = new AsyncLocal<string>();

        public static CacheService CacheService => _connection.Value.Item1;
        public static NotificationService NotificationService => _connection.Value.Item2;
        public static string Agent => NotificationService.Agent;
        public static string Instance => _instance.Value;
        // bool isFunction

        public static void SetConnection(CacheService cacheService, NotificationService notificationService)
        {
            _connection.Value = (cacheService, notificationService);
        }

        public static Task EnterContext(string instance, Func<Task> action)
        {
            return Task.Run(() =>
            {
                _instance.Value = instance;
                return action();
            });
        }

        public static Task<T> EnterContext<T>(string instance, Func<Task<T>> action)
        {
            return Task.Run(() =>
            {
                _instance.Value = instance;
                return action();
            });
        }
    }
}

File: agent/python/perper/model/async_locals.py
from contextvars import ContextVar

connection = ContextVar("connection")
instance = ContextVar("instance")


def set_connection(cache_service, notification_service):
    connection.set((cache_service, notification_service))


def get_cache_service():
    return connection.get()[0]


def get_notification_service():
    return connection.get()[1]


def get_local_agent():
    return get_notification_service().agent


def get_instance():
    return instance.get()[0]


def get_execution():
    return instance.get()[1]


def enter_context(_instance, execution, callback):
    instance.set((_instance, execution))
    return callback()


def set_context(_instance, execution):
    instance.set((_instance, execution))
